Embedding merging

lstein opened the pull request:
# WIP FEAT: Dynamically load and merge multiple embedded models

- The embedding_manager is now enhanced with `ldm.invoke.concepts_lib`, which handles dynamic downloading and caching of embedded models from the Hugging Face concepts library (https://huggingface.co/sd-concepts-library)
 
- Downloading of a embedded model is triggered by the presence of one or more `<concept>` tags in the prompt, where "concept" corresponds to the name of the Hugging Face concept model (e.g. `<poolroom>`)
    
- Once the embedded model is downloaded, its trigger phrase will be loaded into the embedding manager and the prompt's` <concept>` tag will be replaced with the `<trigger_phrase>`
    
- The downloaded model stays on disk for fast loading later. It will be saved in the invokeai run directory under `models`.
    
- The CLI autocomplete will complete partial `<concept>` tags for you. Type a`<` and hit tab to get all ~700 concepts.

# API

The API is handled by the Concepts() object, defined in `ldm.invoke.concepts_lib`. The infamous Generate() object, creates a singleton at startup, and you can access this object with generate.concept_lib(). 
    
## BUGS AND LIMITATIONS:
    
- MODEL NAME VS TRIGGER PHRASE
   You must use the name of the concept embed model from the SD  library, and not the trigger phrase itself. 
    Usually these are the same, but not always. For example, the model named "hoi4-leaders" corresponds to the 
    trigger `<HOI4-Leader>`
    
   One reason for this design choice is that there is no apparent constraint on the uniqueness of the trigger 
   phrases in the Hugging Face library, and one trigger  phrase may map onto multiple models. So we use the 
   model name to maximize reproducibility.
    
   The second reason is that there is no way I know of to search Hugging Face for models with certain trigger 
   phrases. So we'd have to download all 700 models to index the phrases.
    
   The problem this presents is that this may confuse users, who will want to reuse prompts from distributions 
   that use the trigger phrase directly. Usually this will work, but not always.

   It's also slightly confusing code because <concepts> are replaced with <triggers> in the prompt before it gets 
   sent to `txt2img`, and then the <triggers> are replaced with <concepts> before the metadata gets written out, 
   so that the prompts can be run through again and return the same results.
    
- WON'T WORK ON A FIREWALLED SYSTEM
    If the host running IAI has no internet connection, it can't
    download the concept libraries. I will add a script that allows
    users to preload a list of concept models.
    
- BUG IN PROMPT REPLACEMENT WHEN MODEL NOT FOUND
    There's a small bug that occurs when the user provides an invalid model name. The <concept> gets replaced with <None> in the prompt.

## Demo
```
invoke> wallpaper <fractal-flame>
>> Downloading sd-concepts-library/neon-pastel......4.84Kb
>> Current embedding manager terms: *, <moebius>, <garcon-the-cat>, <poolrooms>, <dark-forest>, <cancer_style>, <galaxy-explorer>, <fractal-flame>
etc
```
![000754 44865809](https://user-images.githubusercontent.com/111189/202982978-061976af-895c-471e-8c55-542bcd866630.png)

```invoke> wallpaper <neon-pastel>
concept neon-pastel is not downloaded
>> Downloading sd-concepts-library/neon-pastel......4.84Kb
>> Current embedding manager terms: *, <moebius>, <garcon-the-cat>, <poolrooms>, <dark-forest>, <cancer_style>, <galaxy-explorer>, <fractal-flame>, <neon-pastel>
Generating:   0%|                                                                                                                                                   | 0/1 [00:00<?, ?it/s]>> Ksampler using model noise schedule (steps >= 30)
etc
```
![000755 1305341537](https://user-images.githubusercontent.com/111189/202982990-67e7d0dc-a2de-4c46-bf76-b36d744383af.png)

## Multiple triggers in one prompt?
  I am not convinced that adding multiple triggers to the same prompt does anything coherent. Sometimes it creates a coherent merging of the styles, and sometimes not. Combining a subject with a style, however, does seem to do the expected thing. For example `my cat <garcon-the-cat> in the style <line-art>`. Garcon is a little black cat with a white bib, and sure enough you get a line-art image of that cat (or at least one image - don't know who that gray cat is):

![000757 4177530289](https://user-images.githubusercontent.com/111189/202983829-44997d2f-097c-46f1-b29b-f83fc25f73d5.png)

psychedelicious commented:
Lincoln, would you mind giving me the ELI5-level description of what this actually does? I'm lost! (ELI5 = explain like I'm 5 😁 )

lstein commented:
Hah. That's what happens when I'm composing a PR late at night.

Hugging Face has created a library of Textual Inversions that have been contributed by the community: https://huggingface.co/sd-concepts-library

Each textual inversion adds a subject (like "my cat"), a style ("string art") or an artist ("picasso") to the standard stable diffusion model. Each one is a really small file, typically a few kB, and you can add multiple ones to the same base model at run time. Each textual inversion has a trigger phrase that invokes it, and by convention, they use the syntax `<phrase>`.

So I can get a picture in string art style by entering a prompt like `a picture of a hearty breakfast <string_art>`

Technically, the textual inversion files are called "embeds" and when you load multiple ones it is "embedding merging".

What I did is to write some code which scans the prompt for trigger phrases that look like `<trigger_phrase>`, then connect to Hugging Face to see if this is one of the 700-odd textual inversions. If it is, InvokeAI downloads the file, installs it, and merges it into the current Stable Diffusion model. It then generates the requested image. From then on, this trigger phrase will be available without re-downloading.

The confusing thing is that sometimes the name of the TI on Hugging Face is the same as the trigger phrase, and sometimes it is a bit different. For example, the TI model named `hoi4-leaders` is triggered by `<HOI4-Leader>`. Unfortunately I can't look up the trigger phrases for each of the 700 models without downloading them all and opening up 700 files. So the compromise is that you trigger with the name of the model (`theodore roosevelt <hoi4-leaders>`) rather than the true trigger phrase (`theodore roosevelt <HOI4-Leader>`). But usually they're the same and I think users will not mind?

damian0815 commented on ldm/invoke/CLI.py:
```diff
@@ -487,6 +492,19 @@ def do_command(command:str, gen, opt:Args, completer) -> tuple:
     return command, operation
 
+def concepts_to_triggers(gen, prompt:str)->str:
+    concepts = re.findall('<([^>]+)>',prompt)
```
this regex causes a huge security hole

`invoke> "a <my arbitrary set of words here>" -s 2 -t`

causes an attempt to build a URL with path `/sd-concepts-library/big thing with a big gap/resolve/main/README.md`

the regex should be far more restrictive (it should probably only match `A-Za-Z0-0_\-`, explicitly typed out like that)

also it should be a constant to avoid copy/pasting.

lstein replied:
There are actually two very similar-looking but different regexps here, but point taken. I've restricted the inner pattern to [\w-] (which should be equivalent to the pattern you recommended), and store the precompiled patterns into attributes of the object at init time.

damian0815 commented on ldm/invoke/CLI.py:
```diff
@@ -213,7 +214,8 @@ def main_loop(gen, opt):
 
         # try to relativize pathnames
-        for attr in ('init_img','init_mask','init_color','embedding_path'):
+        #        for attr in ('init_img','init_mask','init_color','embedding_path'):
```
either `embedding_path` should be put back or the reference to it below should be deleted

lstein replied:
The reference is now deleted. This section was relativising paths to the output directory, which makes sense for images but not for the embedding path. I had commented the reference to embedding path just in case there was some reason for this that would show up during testing, but no, just a past mistake.

damian0815 commented on ldm/invoke/args.py:
```diff
@@ -547,6 +547,7 @@ def _create_arg_parser(self):
         )
         render_group.add_argument(
```
(this reference here)

lstein replied:
Could you explain what you mean here?

damian0815 replied:
i was referring to the previous comment "either embedding_path should be put back or the reference to it below should be deleted"

damian0815 replied:
as-is the --embedding_path arg will still be parsed and also reported by -h, but the implementation is ignoring it

lstein replied:
In latest version the directory or file pointed to by `--embedding_path` will be scanned for embedding files and loaded. Default is `~/invokeai/embeddings`. Embedding loading can be disabled using `--no-embedding`

damian0815 commented on ldm/invoke/concepts_lib.py:
```diff
@@ -0,0 +1,143 @@
+        '''
+        def do_replace(match)->str:
+            return self.trigger_to_concept(match.group(1)) or f'<{match.group(1)}>'
+        return re.sub('(<[^>]+>)', do_replace, prompt)
```
duplication of this regex in four separate places over two python source fails breaks both the DRY (Don't Repeat Yourself) principle in two different ways at the same time

lstein replied:
Fixed. I only count the regexp being repeated twice, and in fact the two instances are slightly different (one includes the angle brackets in the match group and the other doesn't). Where are the other two instances?

damian0815 commented on scripts/configure_invokeai.py:
```diff
@@ -594,7 +594,7 @@ def initialize_rootdir(root:str,yes_to_all:bool=False):
     for name in ('models','configs','scripts','frontend/dist'):
         os.makedirs(os.path.join(root,name), exist_ok=True)
-    for src in ('configs','scripts','frontend/dist'):
+    for src in ('configs','frontend/dist'):
```
should this be here? seems unrelated

lstein replied:
Shouldn't be in this PR. Will remove.

victorca25 commented:
@lstein the token phrases are in the dict object of the textual inversion model. If you look at the keys of the loaded model, you'll see the keys, so you avoid guessing them

lstein commented:
The regexp is actually slightly different in two of those places. Note the
relative placing of the '<' and '(' symbols.

Will replace the inner part with [\w-]. Thanks for catching the security
hole.

On Mon, Nov 21, 2022 at 1:50 PM Damian Stewart ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In ldm/invoke/concepts_lib.py
> <https://github.com/invoke-ai/InvokeAI/pull/1526#discussion_r1028412535>:
>
> > +        '''
> +        concept = self.concept_names.get(trigger,None)
> +        return f'<{concept}>' if concept else f'{trigger}'
> +
> +    def replace_triggers_with_concepts(self, prompt:str)->str:
> +        '''
> +        Given a prompt string that contains <trigger> tags, replace these
> +        tags with the concept name. The reason for this is so that the
> +        concept names get stored in the prompt metadata. There is no
> +        controlling of colliding triggers in the SD library, so it is
> +        better to store the concept name (unique) than the concept trigger
> +        (not necessarily unique!)
> +        '''
> +        def do_replace(match)->str:
> +            return self.trigger_to_concept(match.group(1)) or f'<{match.group(1)}>'
> +        return re.sub('(<[^>]+>)', do_replace, prompt)
>
> duplication of this regex in four separate places over two python source
> fails breaks both the DRY (Don't Repeat Yourself) principle in two
> different ways at the same time
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/invoke-ai/InvokeAI/pull/1526#pullrequestreview-1188789525>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA3EVMP4R7TEOVPLJYVXHDWJO76DANCNFSM6AAAAAASGJVDZ4>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>


-- 

*Lincoln Stein*

Head, Adaptive Oncology, OICR

Senior Principal Investigator, OICR

Professor, Department of Molecular Genetics, University of Toronto



Tel: 416-673-8514

Cell: 416-817-8240

***@***.***



*E**xecutive Assistant*

Michelle Xin

Tel: 647-260-7927

***@***.*** ***@***.***>*



*Ontario Institute for Cancer Research*

MaRS Centre, 661 University Avenue, Suite 510, Toronto, Ontario, Canada M5G
0A3

@OICR_news
<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Ftwitter.com%2Foicr_news&data=04%7C01%7CMichelle.Xin%40oicr.on.ca%7C9fa8636ff38b4a60ff5a08d926dd2113%7C9df949f8a6eb419d9caa1f8c83db674f%7C0%7C0%7C637583553462287559%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=PS9KzggzFoecbbt%2BZQyhkWkQo9D0hHiiujsbP7Idv4s%3D&reserved=0>
 | www.oicr.on.ca



*Collaborate. Translate. Change lives.*



This message and any attachments may contain confidential and/or privileged
information for the sole use of the intended recipient. Any review or
distribution by anyone other than the person for whom it was originally
intended is strictly prohibited. If you have received this message in
error, please contact the sender and delete all copies. Opinions,
conclusions or other information contained in this message may not be that
of the organization.

hipsterusername commented:
Huggingface 'integration' seems to be less important for (pro) users than the ability to use >1 embed.

Aside from downloading from HF and merging, what would the commands/process be for a user to initialize with >1 embedding_path? (I didn't see this called out in the docs, and haven't looked at the code to answer my own question 🙃 )

lstein commented:
> Huggingface 'integration' seems to be less important for (pro) users than the ability to use >1 embed.
> 
> Aside from downloading from HF and merging, what would the commands/process be for a user to initialize with >1 embedding_path? (I didn't see this called out in the docs, and haven't looked at the code to answer my own question upside_down_face )

`invoke.py --embedding some_directory/*.pt`

Just pass a list of `.pt` or `.bin` files to the `--embedding` argument, and they will all be loaded. I'm modifying the code slightly now so that if you pass the path to a directory, the directory will be scanned for .bin and .pt files and they'll each be loaded.

lstein commented:
> @lstein the phrases are in the dict object of the textual inversion model. If you look at the keys of the loaded model, you'll see the phrase, so you avoid guessing them

I don't guess them. I get them out of the model, just as you suggest. The issue is that there are 800 models up on Hugging Face and to get all the phrases I would have to download them all and either load the model or read the `token_identifier` file. For this reason, the user enters `<concept-name>` and not `<token-identifier>`. They are usually the same, and this also avoids the problem of two unrelated concepts having chosen the same identifier.

lstein commented:
@damian0815 I've finished what I wanted to for the time being. I still have to test on a machine that is not connected to the Internet and probably provide a concept download script for this scenario, but I think it's ready for a merge. I'm taking it out of draft state and will look for your final review.

hipsterusername commented:
> pass the path to a directory, the directory will be scanned for .bin and .pt files and they'll each be loaded.

I would assume that this directory should be the same place that HF .bin and .py files are being downloaded to? Would it be simpler to just update to do the following:
- Make embedding imports an opt-out
- Have an embed folder (default, empty)
- Users can download or put embed files there if they want them loaded
- Users can exclude embeds when initializing

Random though at 11th hour, just something to think about for UX.

damian0815 commented:
> Make embedding imports an opt-out

i have added code that refuses to overwrite user-defined embeds (i.e. ones loaded via the `--embeddings` command line argument) with automagically loaded huggingface embeds, which probably makes this unnecessary.

> Have an embed folder (default, empty)
> Users can download or put embed files there if they want them loaded

agreed good feature, but should go in a new PR. also being able to drag and drop on to the web UI would be good (same with ckpt files)

> Users can exclude embeds when initializing

oof can of worms, hah.

damian0815 commented:
above comment is @hipsterusername

damian0815 commented on configs/sd-concepts.txt:
```diff
@@ -0,0 +1,803 @@
+sd-concepts-library/001glitch-core
```
this will fail with the \w regex

(this would actually be a good opportunity to make a unit test that pushes all these strings through your regex and only passes if they match; then you can make the regex as restrictive as possible for security reasons while ensuring it still does what it needs)

damian0815 commented:
> The regexp is actually slightly different in two of those places. Note the relative placing of the '<' and '(' symbols.

all the more reason to make it a constant. it's much easier to notice the difference between these two terms:

```
        self.match_trigger = re.compile('(<[\w\-]+>)')
        self.match_concept = re.compile('<([\w\-]+)>')
```

than between these two regexp:
```
        '(<[\w\-]+>)'
        '<([\w\-]+)>'
```

(see this is why a literary studies MA makes sense alongside a BSc in software engineering ;) )

lstein commented:
> > pass the path to a directory, the directory will be scanned for .bin and .pt files and they'll each be loaded.
> 
> I would assume that this directory should be the same place that HF .bin and .py files are being downloaded to? Would it be simpler to just update to do the following:
> 
> * Make embedding imports an opt-out
> * Have an embed folder (default, empty)
> * Users can download or put embed files there if they want them loaded
> * Users can exclude embeds when initializing
> 
> Random though at 11th hour, just something to think about for UX.

That sounds good. I'll implement an embed folder concept.

lstein commented:
> > The regexp is actually slightly different in two of those places. Note the relative placing of the '<' and '(' symbols.
> 
> all the more reason to make it a constant. it's much easier to notice the difference between these two terms:
> 
> ```
>         match_trigger
>         match_concept
> ```
> 
> than between these two regexp:
> 
> ```
>         '(<[\w\-]+>)'
>         '<([\w\-]+)>'
> ```
> 
> and also note how the intention is clearer? i can read `match_trigger` and `match_concept` and know that you wrote it with the intention to match a trigger, vs matching a concept. whereas `(<[\w\-]+>)` is just babble.
> 
> (see this is why a literary studies MA makes sense alongside a BSc in software engineering ;) )

No arguments at all here!

damian0815 reviewed (approved):
LGTM

psychedelicious commented:
Tagging issue https://github.com/invoke-ai/InvokeAI/issues/865

damian0815 commented:
> Tagging issue #865

different kind of model merging AFAICT

hipsterusername commented:
> Tagging issue #865

Aye - Believe that is referencing Models, not embeds; although @lstein is looking at model merging as well.

psychedelicious reviewed (changes requested):
I have done a brief test - pretty cool feature.

Three changes:

When using an incorrect embedding trigger token (generation still proceeds):
`urllib.error.HTTPError: HTTP Error 404: Not Found`

So I think a caught `HTTPError` should result in a warning message that the token wasn't found on HF but that we will generate anyways.

---

When no internet connection:
`urllib.error.URLError: <urlopen error [Errno 8] nodename nor servname provided, or not known>`

Similarly, a caught `URLError` should result in a warning that there was a network connectivity issue, but we will generate anyways. 

---

I think the autocomplete should show the trailing caret `>`:
```
invoke> <pool
<pool-test  <poolrooms
```

n00mkrad commented:
I was asking for model interpolation, not Textual Inversion embeddings

lstein commented:
Implemented.

> > > pass the path to a directory, the directory will be scanned for .bin and .pt files and they'll each be loaded.
> > 
> > 
> > I would assume that this directory should be the same place that HF .bin and .py files are being downloaded to? Would it be simpler to just update to do the following:
> > 
> > * Make embedding imports an opt-out
> > * Have an embed folder (default, empty)
> > * Users can download or put embed files there if they want them loaded
> > * Users can exclude embeds when initializing
> > 
> > Random though at 11th hour, just something to think about for UX.
> 
> That sounds good. I'll implement an embed folder concept.

Done.

lstein commented:
> I have done a brief test - pretty cool feature.
> 
> Three changes:
> 
> When using an incorrect embedding trigger token (generation still proceeds): `urllib.error.HTTPError: HTTP Error 404: Not Found`
> 
> So I think a caught `HTTPError` should result in a warning message that the token wasn't found on HF but that we will generate anyways.
> 
> When no internet connection: `urllib.error.URLError: <urlopen error [Errno 8] nodename nor servname provided, or not known>`
> 
> Similarly, a caught `URLError` should result in a warning that there was a network connectivity issue, but we will generate anyways.

This is now done.

> 
> I think the autocomplete should show the trailing caret `>`:

This will happen when there is only one valid completion.

psychedelicious commented:
If I recall correctly, I didn't get the trailing caret ever on the autocomplete results. 

This may be due to readline differences on macOS. 

Only a potential issue because it gives the impression that leaving the trailing caret off is correct.

lstein commented:
I've addressed the comments in the code review.

psychedelicious reviewed (changes requested):
1. I had issues getting `invoke.py` to run on this branch:

I get an alert saying I have misconfigured model files. 

```
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
   You appear to have a missing or misconfigured model file(s).
   The script will now exit and run configure_invokeai.py to help fix the problem.
   After reconfiguration is done, please relaunch invoke.py.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

I continue the `configure_invokeai.py` script, just taking the default options (which appear to be correct) and it appears to be successful, but when I try again, it fails the same way.

I see my `.invokeai` file has no default root or output directory set:

```
❯ cat ~/.invokeai
# InvokeAI initialization file
# Put frequently-used startup commands here, one or more per line
# Examples:
# --web --host=0.0.0.0
# --steps 20
# -Ak_euler_a -C10.0
```

I tried deleting this and it recreate the same exact file again.

Looks like my root and outdir are not being written to the file. I manually added them and now it works.

2. `URLError` doesn't appear to be caught (or caught in the right place), invoke.py crashes when I disconnect my internet:
`AttributeError: 'URLError' object has no attribute 'code'`

3. The autocomplete list of valid values do not include the trailing caret still, which is kinda confusing. If you just manually type in what autocomplete shows you, instead of pressing tab, it won't work of course bc you are missing the trailing caret.

Here's what my autocomplete looks like:
```bash
invoke> <poo
<pool-test  <poolrooms
invoke> <pool
<pool-test  <poolrooms
invoke> <poolrooms> -s 5
```

lstein commented:
I had the same issues last night while working on an unrelated branch. It was due to the branch not being up to date with the recent changes to `development`. I have updated the PR with current development and everything seems to be working ok.

I've fixed the autocomplete so that the ending '>' is now displayed. @psychedelicious when you have a chance, please test again. I appreciate it.


> 1. I had issues getting `invoke.py` to run on this branch:
> 
> I get an alert saying I have misconfigured model files.
> 
> ```
> !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
>    You appear to have a missing or misconfigured model file(s).
>    The script will now exit and run configure_invokeai.py to help fix the problem.
>    After reconfiguration is done, please relaunch invoke.py.
> !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
> ```
> 
> I continue the `configure_invokeai.py` script, just taking the default options (which appear to be correct) and it appears to be successful, but when I try again, it fails the same way.
> 
> I see my `.invokeai` file has no default root or output directory set:
> 
> ```
> ❯ cat ~/.invokeai
> # InvokeAI initialization file
> # Put frequently-used startup commands here, one or more per line
> # Examples:
> # --web --host=0.0.0.0
> # --steps 20
> # -Ak_euler_a -C10.0
> ```
> 
> I tried deleting this and it recreate the same exact file again.
> 
> Looks like my root and outdir are not being written to the file. I manually added them and now it works.
> 
> 2. `URLError` doesn't appear to be caught (or caught in the right place), invoke.py crashes when I disconnect my internet:
>    `AttributeError: 'URLError' object has no attribute 'code'`
> 3. The autocomplete list of valid values do not include the trailing caret still, which is kinda confusing. If you just manually type in what autocomplete shows you, instead of pressing tab, it won't work of course bc you are missing the trailing caret.
> 
> Here's what my autocomplete looks like:
> 
> ```shell
> invoke> <poo
> <pool-test  <poolrooms
> invoke> <pool
> <pool-test  <poolrooms
> invoke> <poolrooms> -s 5
> ```

psychedelicious commented:
1. Autocomplete carets are shown now :)

2. `URLError` still not caught when no internet connection; invoke.py crashes :(

3. Still the `.invokeai` file is not being created correctly - but I found the issue!

`invoke.py` calls `ldm.invoke.CLI.main()`, which immediately instantiates `Args()`.

`Args()` checks if `~/.invokeai exists and if not, and if not, it makes one in `Args._create_init_file()`. *This function creates the file with the "old" style/content of `.invokeai`.*

After `Args()` does its thing, flow goes back into `ldm.invoke.CLI.main()`, which sees the `.invokeai` file but doesn't see the `root` and `outdir` flags in it, recognizing it is incorrect. It then fires up `configure_invokeai.py`.

In that script, there is no facility to fix a broken or improperly configured `.invokeai`. E.g `configure_invokeai.py` L605, it conditionally creates the new file with the new format/content: `    if not os.path.exists(init_file):`

This of course doesn't happen bc I already have an `.invokeai` file. So basically unless your `.invokeai` file is not already correct, the script can never correct it.

lstein commented:
Thanks for tracking down the issue with .invokeai. I was trying to reproduce the issue by removing the file and running `configure_invokeai.py`, but now I see you triggered it using a different path. One more try at this!!!!

lstein commented:
@psychedelicious I've fixed both the problems -- I hope for real this time. Thank you for your help tracking them down.

psychedelicious commented:
No worries, happy to help. Not quite there yet but very close!:

1. If I run `invoke.py` with no network connection at all, I get a URLError. I had previously tested by getting to `invoke>` prompt with network connectivity and *then* disconnecting from network. This URLError is coming from patchmatch:

```python
│ /Users/spencer/mambaforge/envs/invokeai/lib/python3.10/site-packages/patchmatch/patch_match.py:1 │
│ 04 in <module>                                                                                   │
│                                                                                                  │
│   101 import platform                                                                            │
│   102                                                                                            │
│   103 # Get release information from github                                                      │
│ ❱ 104 release_response = urlopen(release_url)                                                    │
│   105 release_json = json.loads(release_response.read())                                         │
│   106                                                                                            │
│   107 # Filter to assets for platform                                                            │
│
```

I suppose the whole shebang may need to be wrapped with a try...

2. If I start with connectivity, then disconnect, and try to use an embedding which was not previously loaded, this is handled gracefully but I think there is a misplaced quote or something... note the `str(e)`:

`>> Downloading sd-concepts-library/amogus...str(e). This may reflect a network issue. Generation will continue without the concept.`

damian0815 reviewed (approved):
code looks good but i didn't test it

lstein commented:
> No worries, happy to help. Not quite there yet but very close! I suppose this is difficult to test bc you cannot disconnect from the network due to your setup :/
> 
> 1. If I run `invoke.py` with no network connection at all, I get a URLError. I had previously tested by getting to `invoke>` prompt with network connectivity and _then_ disconnecting from network. This URLError is coming from patchmatch:
> 
> ```python
> │ /Users/spencer/mambaforge/envs/invokeai/lib/python3.10/site-packages/patchmatch/patch_match.py:1 │
> │ 04 in <module>                                                                                   │
> │                                                                                                  │
> │   101 import platform                                                                            │
> │   102                                                                                            │
> │   103 # Get release information from github                                                      │
> │ ❱ 104 release_response = urlopen(release_url)                                                    │
> │   105 release_json = json.loads(release_response.read())                                         │
> │   106                                                                                            │
> │   107 # Filter to assets for platform                                                            │
> │
> ```
> 
> I suppose the whole shebang may need to be wrapped with a try...

I'll fix this in a separate PR.

> 2. If I start with connectivity, then disconnect, and try to use an embedding which was not previously loaded, this is handled gracefully but I think there is a misplaced quote or something... note the `str(e)`:
> 
> `>> Downloading sd-concepts-library/amogus...str(e). This may reflect a network issue. Generation will continue without the concept.`

The string formatting issue is now fixed.

psychedelicious reviewed (approved):
Everything working as it should (minus the patchmatch issue to be addressed separately). Thanks!
